### PR TITLE
 Fix display of category from other shop 

### LINKED ISSUE #31797: Disallow display of category from other shop
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

With a multishop installation, you can display the list of products of a category from another store, even if this category is not associated with the current store.  
Just need to know the category ID
ex: https://shop.com/1234-id-cat-other-shop

### Expected behavior

display a 404 page if the category is not associated with the current shop

### Steps to reproduce

1) you have a multisite installation with two shop (ore more)
2) in general context create a category tree 
- root
-- shop 1
--- cat-1
-- shop 2
--- cat-2
3) got to shop 1 assign cat-1 to this shop  
4) got to shop 2 assign cat-2 to this shop  

Get the id of cat-1
Got to front of shop 2, on listing of cat-2, 
change the id of cat-2 with id cat-1   
You will see category of shop-1 on shop-2

### PrestaShop version(s) where the bug happened

1.7.8.8

### PHP version(s) where the bug happened

7.4

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by mkdgs**
Display a 404 error page when another shop's category is called see #31797 was PR #31798 

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           |  8.0.x
| Description?      | Respect multi-store category activity state
| Type?             | bug fix
| Category?         | FO
| BC breaks?        |  no
| Deprecations?     |  no
| How to test?      | see issue related
| Fixed ticket?     | Closes #31797
| Related PRs       | 
| Sponsor company   |

### CODE REVIEW

**Review by kpodemski — DISMISSED** [review]

**Review by SharakPL — DISMISSED** [review]

**Review by florine2623 — COMMENTED** [review]
Hello @mkdgs ,

Thanks for the PR !

Here's what I have :
Screenshot 2023-03-23 at 14 41 25


Multiple warnings, and I still have access to the category even though in BO the related shop is not checked.

Screenshot 2023-03-23 at 14 42 28

Could you check ? ^^
Thanks!

**Comment by mkdgs** [comment]
@florine2623 this fix was initially made for 1.7.x the comportment seems different.
I've added a reset for category, that's fixed it.

Let me know if you see anything else :)

**Comment by kpodemski** [comment]
Hi @mkdgs 

I don't know if getting an empty category instance is a good idea to solve the problem. It's more like a workaround. I think the problem is elsewhere, if the category shouldn't be displayed, why are there products displayed on the screenshot? It should return a 404 page.

**Comment by mkdgs** [comment]
@kpodemski, i think is the best think to do, because this category don't exist on the shop and has no reasons to be set here. 
The products displayed was the products on this category from other shop, and they are not allowed to be seen on this shop.
They was displaying because previous Category instance stay defined, even if a error page.
So we need to reset this instance, if we have no category we have no products.

This appear to be the simplest and most effective way to get a correct 404 page :)

**Review by matks — CHANGES_REQUESTED** [review]
> I don't know if getting an empty category instance is a good idea to solve the problem. It's more like a workaround

I agree with @kpodemski : creating a new empty object just to make the page work is a workaround and I think it will bite us back in a few weeks/months 😅 

I think we need to find a better way to have a 404 page 😞

**Comment by mkdgs** [comment]
damn ....  this issue will remain open until the death of prestashop.
@matks It's not just making a new instance, it's resetting a previous instance.

This instance is already set with a wrong value !!!
This not a problem in weeks/months it's already a problem right now.

have you looked at the code ?
```php
public function init()
    {
        $id_category = (int) Tools::getValue('id_category');
        // this category is always set here !
        $this->category = new Category(
            $id_category,
            $this->context->language->id
        );

        parent::init();
        
        if (!Validate::isLoadedObject($this->category) || !$this->category->active || !$this->category->existsInShop($this->context->shop->id)) {
            header('HTTP/1.1 404 Not Found');
            header('Status: 404 Not Found');
            $this->setTemplate('errors/404');
            // there is no category available, but the instance is already defined. So we have to reset it ... 
            $this->category = new Category();
            $this->notFound = true;

**Comment by kpodemski** [comment]
@mkdgs thing is that this instance should not be used at all on 404 page, right? if the category is inactive in the multi-store, we should be redirected to 404 page where nothing from the `Category` is required, am I right?

I'll investigate this issue today.

File: controllers/front/listing/CategoryController.php [hunk 1]
```diff
@@ -80,10 +80,11 @@ public function init()
 
         parent::init();
 
-        if (!Validate::isLoadedObject($this->category) || !$this->category->active) {
+        if (!Validate::isLoadedObject($this->category) || !$this->category->active || !$this->category->existsInShop($this->context->shop->id)) {
             header('HTTP/1.1 404 Not Found');
             header('Status: 404 Not Found');
             $this->setTemplate('errors/404');
+            $this->category = new Category();
```
Comment (kpodemski) [inline controllers/front/listing/CategoryController.php:87]:
```suggestion
```

**Review by kpodemski — CHANGES_REQUESTED** [review]
In the `initContent` you need to add:

`&& $this->category->existsInShop($this->context->shop->id)` after:

`&& $this->category->checkAccess($this->context->customer->id)`

and it will work without hacky way of instantiating Category :)

**Comment by mkdgs** [comment]
@kpodemski  the category is already instantiated at this step. If you not reset it will stay instantiated with the wrong value.
I think checkAccess() must return a 403 error code, not a 404 
Not tested but i don't see how this can resolve the issue. 

Another "better" way to do it  is to set $this->category only if everything is OK.
But let $this->category with anything else other than an empty instance may throw error later (like for breadcrumbs) 
eg:

```php
public function init()
    {
        $id_category = (int) Tools::getValue('id_category');
        // this category is always set here !
        $tmp_category = new Category(
            $id_category,
            $this->context->language->id
        );

        parent::init();
        
        if (!Validate::isLoadedObject($tmp_category) || !$tmp_category->active || !$tmp_category->existsInShop($this->context->shop->id)) {
            header('HTTP/1.1 404 Not Found');
            header('Status: 404 Not Found');
            $this->setTemplate('errors/404');
            $this->notFound = true;
       }
       // check access...

// all is fine 
        $this->category = $tmp_category;

**Comment by kpodemski** [comment]
@mkdgs if you do what I suggested we don't need to look for workarounds, we have to have this extra check in the initContent, otherwise the template tries to use category vars which should not be the case

**Comment by mkdgs** [comment]
@kpodemski i got it, i misread and not see it's about init**Content**
But why not:
```php
 public function initContent()
    {
        parent::initContent();

        if (
            !$this->notFound
            && $this->category->checkAccess($this->context->customer->id)
        ) {
            $this->doProductSearch(
                'catalog/listing/category',
                [
                    'entity' => 'category',
                    'id' => $this->category->id,
                ]
            );
        }
    }
```
and what is the use of the properties $customer_access ? 
we can set it and use it ?

And i still think it's safer to reset $this->category, at least to NULL, both for the two check of existing and access

**Comment by kpodemski** [comment]
@mkdgs I have no idea what is `$customer_access`, it seems like something abandoned

I'd suggest keeping this code:

```
public function initContent()
    {
        parent::initContent();

        if (
            Validate::isLoadedObject($this->category)
            && $this->category->active
            && $this->category->checkAccess($this->context->customer->id)
            && $this->category->existsInShop($this->context->shop->id)
        ) {
            $this->doProductSearch(
                'catalog/listing/category',
                [
                    'entity' => 'category',
                    'id' => $this->category->id,
                ]
            );
        }
    }
```

for now

**Comment by kpodemski** [comment]
I agree that it's room for improvement there, but as this is quite a nasty bug, we should keep the simple way of fixing it.

**Comment by kpodemski** [comment]
hi @mkdgs 👋🏻 

do you think you could finish the PR or shall I takeover?

**Comment by matks** [comment]
Hi,

Since there was no activity since April, I will follow [our guidelines](https://devdocs.prestashop.com/1.7/project/maintainers-guide/reviewing-pull-requests/#stale-pull-requests) and I'll close this pull request as it seems stale. Feel free to reopen or open another one if you think it's still relevant.

@kpodemski you can also reopen it if you wish to finish it 😉 

Thanks!

**Pull request closed without merging**